File: Algorithms/043. Multiply Strings/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def multiply(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        if num1=='0' or num2=='0':
            return '0'
        l1=len(num1)
        l2=len(num2)
        l=l1+l2-1
        ans=[0]*l
        sum=0
        carry=0
        for i in range(1,l+1):
            sum=0
            for a1 in range(i):
                a2=i-a1
                a1+=1
                if a1<=l1 and a2<=l2:
                    sum+=(int(num1[-a1])*int(num2[-a2]))
            sum+=carry
            carry=sum/10
            ans[-i]=sum%10
        res='' if carry==0 else str(carry)
        for a in ans:
            res+=str(a)
        return res

File: Algorithms/043. Multiply Strings/Solution.java
public class Solution {
    public String multiply(String num1, String num2) {
        if (num1.equals("0")||num2.equals("0")) {
            return "0";
        }
        if (num1.length()<num2.length()) {
            return multiply(num2,num1);
        }
        int l1=num1.length();
        int[] n1=new int[l1];
        for (int i = 0; i <l1; i++) {
            n1[l1-i-1]=Integer.parseInt(num1.substring(i, i+1));
        }
        int l2=num2.length();
        int[] n2=new int[l1];
        for (int i = 0; i <l2; i++) {
            n2[l2-i-1]=Integer.parseInt(num2.substring(i, i+1));
        }
        int[] a=new int[l1+l2-1];
        int sum=0;
        int carry=0;
        for (int i = 0; i < a.length; i++) {
            // 每一位乘级相加的结果
            for (int j = Math.max(0,i-l2+1); j <= Math.min(i,l1-1); j++) {
                sum += n1[j] * n2[i-j];
            }
            // 加进位
            sum += carry;
            carry = sum / 10;
            sum%=10;
            a[i]=sum;
            sum = 0;
        }
        StringBuffer sb=new StringBuffer();
        if (carry!=0) {
            sb.append(String.valueOf(carry));
        }
        for (int i = a.length-1; i >=0 ; i--) {
            sb.append(String.valueOf(a[i]));
        }
        return sb.toString();
    }
}